### PR TITLE
Add max_timeout support in retry_policy

### PR DESCRIPTION
**Pull request by xiangyan99**

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/core/azure-core/tests/test_retry_policy.py [hunk 1]
```diff
@@ -159,3 +160,31 @@ def send(self, request, **kwargs):  # type: (PipelineRequest, Any) -> PipelineRe
...
+            time.sleep(5)
+            raise ServiceResponseError('timeout')
+
+    http_request = HttpRequest('GET', 'http://127.0.0.1/')
+    headers = {'Content-Type': "multipart/form-data"}
+    http_request.headers = headers
+    http_retry = RetryPolicy(retry_total=1, max_timeout=10)
+    pipeline = Pipeline(MockTransport(), [http_retry])
+    with pytest.raises(ServiceResponseError):
+        pipeline.run(http_request)
```
Comment (bryevdv) [inline sdk/core/azure-core/tests/test_retry_policy.py]: Wouldn't the test be safer if:

* `retry_total` was larger to make sure it doesn't trigger before the timeout condition
* the exception was checked for the specific text `"timeout"`

File: sdk/core/azure-core/azure/core/pipeline/policies/_retry.py [hunk 2]
```diff
@@ -377,7 +385,7 @@ def update_context(self, context, retry_settings):
         if retry_settings['history']:
             context['history'] = retry_settings['history']
 
-    def send(self, request):
+    def send(self, request):    # pylint: disable=too-many-statements, too-many-branches
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/pipeline/policies/_retry.py]: When I see this pylint warning, and the fact that we duplicated a ton of code from sync/async, I feel we should have a common logic imported from both places instead.
TLDR: Too much duplication of algorithm/logic here to me

File: sdk/core/azure-core/tests/test_retry_policy.py [hunk 3]
```diff
@@ -159,3 +160,31 @@ def send(self, request, **kwargs):  # type: (PipelineRequest, Any) -> PipelineRe
         pipeline = Pipeline(MockTransport(), [http_retry])
         pipeline.run(http_request)
     os.unlink(f.name)
+
+def test_retry_timeout():
```
Comment (lmazuel) [inline sdk/core/azure-core/tests/test_retry_policy.py]: async test?

File: sdk/core/azure-core/tests/test_retry_policy.py [hunk 4]
```diff
@@ -159,3 +160,31 @@ def send(self, request, **kwargs):  # type: (PipelineRequest, Any) -> PipelineRe
...
+            self.count += 1
+            if self.count > 2:
+                assert self.count <= 2
+            time.sleep(5)
+            raise ServiceResponseError('timeout')
+
+    http_request = HttpRequest('GET', 'http://127.0.0.1/')
+    headers = {'Content-Type': "multipart/form-data"}
+    http_request.headers = headers
+    http_retry = RetryPolicy(retry_total=10, timeout=10)
```
Comment (lmazuel) [inline sdk/core/azure-core/tests/test_retry_policy.py]: This test is designed to take 10seconds, I usually try to avoid such big values since eventually if we have too much test like that CI starts to become slow.

I understand time is an important factor here, but I would encourage to make smaller values. If you're total timeout is 1sec, and you retry every 500ms, you still test your behavior but it's 10X faster.
